File: src/types/plugins.ts
import type { ReactNode } from "react";

import type { PluginInternalPage } from "./item";

export type PluginStaticPageRow = {
  label: string;
  value: string;
};

export type PluginStaticPageSection = {
  title: string;
  rows?: PluginStaticPageRow[];
  paragraphs?: string[];
};

export type PluginStaticPageContent = {
  subtitle?: string;
  sections?: PluginStaticPageSection[];
};

export type PluginPageHelpCommand = {
  command: string;
  description: string;
};

export type PluginPageHelpContent = {
  description?: string;
  examples?: string[];
  commands?: PluginPageHelpCommand[];
};

export type PluginPageActionManifest = {
  actionId: string;
  inputPlaceholder?: string;
  examples?: string[];
};

export type PluginPageDefinition = {
  id: PluginInternalPage;
  tabs: PluginPageTab[];
};

export type PluginPageTab =
  | PluginPlaygroundTab
  | PluginConverterTab
  | PluginFormTab;

export type PluginPageTabBase = {
  id: string;
  title?: string;
  titleKey?: string;
  titleFallback?: string;
};

export type PluginPlaygroundTab = PluginPageTabBase & {
  type: "playground";
  action: string;
  inputPlaceholder?: string;
  inputPlaceholderKey?: string;
  inputPlaceholderFallback?: string;
  examples?: string[];
  submitLabel?: string;
};

export type PluginConverterTab = PluginPageTabBase & {
  type: "converter";
  action: string;
  accept?: string | string[];
  multiple?: boolean;
  maxBytes?: number;
  maxFiles?: number;
  outputDirectorySetting?: string;
  description?: string;
  descriptionKey?: string;
  descriptionFallback?: string;
  chooseFileLabel?: string;
  chooseFileLabelKey?: string;
  chooseFileLabelFallback?: string;
  emptyLabel?: string;
  emptyLabelKey?: string;
  emptyLabelFallback?: string;
  convertingLabel?: string;
  convertingLabelKey?: string;
  convertingLabelFallback?: string;
  resultsLabel?: string;
  resultsLabelKey?: string;
  resultsLabelFallback?: string;
  revealLabel?: string;
  revealLabelKey?: string;
  revealLabelFallback?: string;
  clearLabel?: string;
  clearLabelKey?: string;
  clearLabelFallback?: string;
  fileColumnLabel?: string;
  fileColumnLabelKey?: string;
  fileColumnLabelFallback?: string;
  sizeColumnLabel?: string;
  sizeColumnLabelKey?: string;
  sizeColumnLabelFallback?: string;
  pathColumnLabel?: string;
  pathColumnLabelKey?: string;
  pathColumnLabelFallback?: string;
  emptyResultsLabel?: string;
  emptyResultsLabelKey?: string;
  emptyResultsLabelFallback?: string;
};

export type PluginFormTab = PluginPageTabBase & {
  type: "form";
  action: string;
  submitLabel?: string;
  submitLabelKey?: string;
  submitLabelFallback?: string;
  fields?: PluginFormField[];
  result?: Record<string, unknown>;
};

export type PluginFormField = {
  id: string;
  type?: "string" | "number" | "boolean" | "enum";
  label?: string;
  labelKey?: string;
  labelFallback?: string;
  description?: string;
  descriptionKey?: string;
  descriptionFallback?: string;
  default?: unknown;
  min?: number;
  max?: number;
  step?: number;
  control?: "text" | "number" | "slider" | "checkbox" | "select" | "segmented";
  options?: PluginFormFieldOption[];
};

export type PluginFormFieldOption = {
  value: string | number | boolean;
  label?: string;
  labelKey?: string;
  labelFallback?: string;
};

export type PluginI18nDictionary = Record<string, unknown>;

export type PluginI18nManifest =
  | Record<string, PluginI18nDictionary>
  | {
      defaultLocale?: string;
      translations?: Record<string, PluginI18nDictionary>;
    };

export type PluginDiscoveryReport = {
  manifests: GlimpsePlugin[];
  errors: PluginDiscoveryError[];
};

export type PluginDiscoveryError = {
  path: string;
  error: string;
};

export type PluginEntrypoints = {
  main?: string;
  page?: string;
};

export type PluginEntrypointSource = {
  pluginId: string;
  entrypoint: "main" | "page" | string;
  path: string;
  source: string;
};

export type PluginAssetSource = {
  pluginId: string;
  asset: "styles" | string;
  path: string;
  source: string;
};

export type PluginReadmeSource = {
  pluginId: string;
  path: string;
  source: string;
};

export type PluginTrustStatus = {
  pluginId: string;
  trusted: boolean;
  trustRequired: boolean;
  reason?: string | null;
  trustedAt?: string | null;
  manifestFingerprint: string;
  trustedFingerprint?: string | null;
  version: string;
  trustedVersion?: string | null;
  provenance?: PluginInstallProvenance | null;
};

export type PluginInstallProvenance = {
  installSource: "local" | "remote";
  registryUrl?: string | null;
  downloadUrl?: string | null;
  registrySha256?: string | null;
  installedPackageSha256?: string | null;
  installedAt?: string | null;
};

export type PluginInstallResult = {
  pluginId: string;
  installedPath: string;
  replaced: boolean;
  manifest: GlimpsePlugin;
};

export type PluginUninstallResult = {
  pluginId: string;
  removedPath: string;
  removed: boolean;
};

export type PluginInternalPageManifest = {
  id: PluginInternalPage;
  title: string;
  titleKey?: string;
  titleFallback?: string;
  tags?: string[];
  aliases?: string[];
  boost?: number;
  pageAction?: PluginPageActionManifest;
  help?: PluginPageHelpContent;
  staticPage?: PluginStaticPageContent;
  pageDefinition?: PluginPageDefinition;
};

export type PluginActionManifest = {
  id: string;
  title: string;
  titleKey?: string;
  titleFallback?: string;
  description?: string;
  descriptionKey?: string;
  descriptionFallback?: string;
  input?: Record<string, unknown>;
  output?: Record<string, unknown>;
  aliases?: string[];
};

export type PluginViewerManifest = {
  id: string;
  title: string;
  titleKey?: string;
  titleFallback?: string;
  description?: string;
  descriptionKey?: string;
  descriptionFallback?: string;
  extensions?: string[];
};

export type PluginContributions = {
  internalPage?: PluginInternalPageManifest;
  internalPages?: PluginInternalPageManifest[];
  actions?: PluginActionManifest[];
  viewers?: PluginViewerManifest[];
};

export type PluginDependencyManifest = {
  id: string;
  version?: string;
  optional?: boolean;
};

export type PluginDependencies = {
  plugins?: PluginDependencyManifest[];
  npm?: string[];
};

export type PluginFileReadScope = "none" | "active-tab" | "target-group";
export type PluginFileWriteScope =
  | "none"
  | "declared-output-directory"
  | "target-group";

export type PluginCapabilities = {
  files?: {
    read?: PluginFileReadScope;
    write?: PluginFileWriteScope;
  };
};

export type PluginRegistry = {
  schemaVersion: 1;
  plugins: PluginRegistryEntry[];
};

export type PluginRegistryEntry = {
  id: string;
  name: string;
  version: string;
  apiVersion: string;
  downloadUrl: string;
  sha256: string;
  /** Derived from repositoryUrl during registry validation. Raw registry author metadata is ignored. */
  author?: string;
  category?: string;
  description?: string;
  downloadCount?: number;
  releaseDate?: string;
  fileName?: string;
  readmeUrl?: string;
  sourceUrl?: string;
  repositoryUrl?: string;
  homepageUrl?: string;
  supportUrl?: string;
};

export type RemotePluginInstallResult = PluginInstallResult & {
  source: "remote";
};

export type InternalPageContribution = PluginInternalPageManifest & {
  pluginId: string;
  render: () => ReactNode;
};

export type GlimpsePlugin = {
  id: string;
  name: string;
  version: string;
  apiVersion?: string;
  author?: string;
  releaseDate?: string;
  repositoryUrl?: string;
  homepageUrl?: string;
  supportUrl?: string;
  description?: string;
  defaultLocale?: string;
  i18n?: PluginI18nManifest;
  i18nPath?: string;
  page?: string;
  pageDefinition?: PluginPageDefinition;
  enabledByDefault?: boolean;
  entrypoints?: PluginEntrypoints;
  contributes?: PluginContributions;
  dependencies?: PluginDependencies;
  capabilities?: PluginCapabilities;
  settings?: Record<string, unknown>;
  internalPages?: PluginInternalPageManifest[];
  warnings?: string[];
};

export type PluginRegistryItem = GlimpsePlugin & {
  enabled: boolean;
  trusted: boolean;
  trustStatus?: PluginTrustStatus;
};

File: src-tauri/src/models/plugins.rs
//! Plugin manifest models.
//!
//! Plugin manifests are loaded from:
//!
//! ```text
//! <app_data_dir>/plugins/*/manifest.json
//! ```
//!
//! The frontend receives these definitions and decides how to expose supported
//! contributions in the UI.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginManifest {
    pub id: String,
    pub name: String,
    pub version: String,
    pub api_version: Option<String>,
    pub author: Option<String>,
    pub release_date: Option<String>,
    pub repository_url: Option<String>,
    pub homepage_url: Option<String>,
    pub support_url: Option<String>,
    pub description: Option<String>,
    pub default_locale: Option<String>,
    pub i18n: Option<serde_json::Value>,
    pub i18n_path: Option<String>,
    pub page: Option<String>,
    pub page_definition: Option<serde_json::Value>,
    pub enabled_by_default: Option<bool>,
    pub entrypoints: Option<PluginEntrypoints>,
    pub contributes: Option<PluginContributions>,
    pub dependencies: Option<PluginDependencies>,
    pub capabilities: Option<PluginCapabilities>,
    pub settings: Option<serde_json::Value>,
    pub internal_pages: Option<Vec<PluginInternalPageManifest>>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginDiscoveryReport {
    pub manifests: Vec<PluginManifest>,
    pub errors: Vec<PluginDiscoveryError>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginDiscoveryError {
    pub path: String,
    pub error: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginEntrypoints {
    pub main: Option<String>,
    pub page: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RawPluginManifest {
    pub id: String,
    pub name: String,
    pub version: String,
    pub api_version: Option<String>,
    pub author: Option<String>,
    pub release_date: Option<String>,
    pub repository_url: Option<String>,
    pub homepage_url: Option<String>,
    pub support_url: Option<String>,
    pub description: Option<String>,
    pub default_locale: Option<String>,
    pub i18n: Option<serde_json::Value>,
    pub page: Option<String>,
    pub enabled_by_default: Option<bool>,
    pub entrypoints: Option<PluginEntrypoints>,
    pub backend: Option<serde_json::Value>,
    pub contributes: Option<PluginContributions>,
    pub dependencies: Option<PluginDependencies>,
    pub capabilities: Option<PluginCapabilities>,
    pub settings: Option<serde_json::Value>,
    pub internal_pages: Option<Vec<PluginInternalPageManifest>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginContributions {
    pub internal_page: Option<PluginInternalPageManifest>,
    pub internal_pages: Option<Vec<PluginInternalPageManifest>>,
    pub actions: Option<Vec<PluginActionManifest>>,
    pub viewers: Option<Vec<PluginViewerManifest>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginActionManifest {
    pub id: String,
    #[serde(default)]
    pub title: String,
    pub title_key: Option<String>,
    pub title_fallback: Option<String>,
    pub description: Option<String>,
    pub description_key: Option<String>,
    pub description_fallback: Option<String>,
    pub input: Option<serde_json::Value>,
    pub output: Option<serde_json::Value>,
    pub aliases: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginViewerManifest {
    pub id: String,
    #[serde(default)]
    pub title: String,
    pub title_key: Option<String>,
    pub title_fallback: Option<String>,
    pub description: Option<String>,
    pub description_key: Option<String>,
    pub description_fallback: Option<String>,
    pub extensions: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginDependencies {
    pub plugins: Option<Vec<PluginDependencyManifest>>,
    pub npm: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginCapabilities {
    pub files: Option<PluginFileCapabilities>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginFileCapabilities {
    pub read: Option<PluginFileReadScope>,
    pub write: Option<PluginFileWriteScope>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PluginFileReadScope {
    None,
    ActiveTab,
    TargetGroup,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PluginFileWriteScope {
    None,
    DeclaredOutputDirectory,
    TargetGroup,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginDependencyManifest {
    pub id: String,
    pub version: Option<String>,
    pub optional: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginEntrypointSource {
    pub plugin_id: String,
    pub entrypoint: String,
    pub path: String,
    pub source: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginAssetSource {
    pub plugin_id: String,
    pub asset: String,
    pub path: String,
    pub source: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginReadmeSource {
    pub plugin_id: String,
    pub path: String,
    pub source: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginTrustStatus {
    pub plugin_id: String,
    pub trusted: bool,
    pub trust_required: bool,
    pub reason: Option<String>,
    pub trusted_at: Option<String>,
    pub manifest_fingerprint: String,
    pub trusted_fingerprint: Option<String>,
    pub version: String,
    pub trusted_version: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginInstallResult {
    pub plugin_id: String,
    pub installed_path: String,
    pub replaced: bool,
    pub manifest: PluginManifest,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginUninstallResult {
    pub plugin_id: String,
    pub removed_path: String,
    pub removed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginInternalPageManifest {
    pub id: String,
    #[serde(default)]
    pub title: String,
    pub title_key: Option<String>,
    pub title_fallback: Option<String>,
    pub tags: Option<Vec<String>>,
    pub aliases: Option<Vec<String>>,
    pub boost: Option<f64>,
    pub page_action: Option<PluginPageActionManifest>,
    pub help: Option<PluginPageHelpContent>,
    pub static_page: Option<PluginStaticPageContent>,
    pub page_definition: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginPageActionManifest {
    pub action_id: String,
    pub input_placeholder: Option<String>,
    pub examples: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginPageHelpContent {
    pub description: Option<String>,
    pub examples: Option<Vec<String>>,
    pub commands: Option<Vec<PluginPageHelpCommand>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginPageHelpCommand {
    pub command: String,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginStaticPageContent {
    pub subtitle: Option<String>,
    pub sections: Option<Vec<PluginStaticPageSection>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginStaticPageSection {
    pub title: String,
    pub rows: Option<Vec<PluginStaticPageRow>>,
    pub paragraphs: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginStaticPageRow {
    pub label: String,
    pub value: String,
}
